### PR TITLE
Notification Settings

### LINKED ISSUE #823: Notifications logic should handle case where the user agrees but cancels the iOS dialog.
When we use the Firebase `requestNotificationPermissions` method to request permission we have to guard against the possibility that the user could (intentionally or accidentally) cancel the request.  We should decide how we wish to handle this - e.g. ask again next time?  - and implement accordingly using the OS feedback from `onIosSettingsRegistered` which gives us the actual value.

### LINKED ISSUE #902: Push Notifications - disable push notifications when user changes iOS settings
We need to finish up the TODO over here: https://github.com/WorldHealthOrganization/app/blob/1d03b2194f482a3e3e604faa9e10305d65831f3f/client/flutter/lib/main.dart#L101-L106

Here, we should modify the user's opt-in settings when they change their push notification settings through iOS settings.

### LINKED ISSUE #931: Refactor Push Notification features into a utility class.
**Please follow this checklist. Please check each appropriate box (put an 'x' or check it after creating the issue).**
- [X] Searched the [existing issues](https://github.com/WorldHealthOrganization/app/issues) to ensure you are not creating a duplicate.
- [X] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md).

The code that manages push notification settings should be factored out into a utility class. Currently, the logic is spread throughout multiple views and other classes.

### LINKED ISSUE #936: If the user denies notifications in onboarding there should be a way to enable them later.
If the user denies notifications in onboarding there should be a way in our Settings page to re-enable them.   This can be pretty simple and just show the notifications page again.  However there is a complication on iOS where iOS will only prompt with the system dialog once and after that we may need to redirect the user to settings to enable them manually.  I"m not sure if Firebase gives us any help with this.

### PR DESCRIPTION
**Pull request by kassim**
**Please follow this checklist. Please check each appropriate box (put an 'x' or check it after creating the PR).**
- [x] REQUIRED: Do you have an Issue **assigned to you** within the v1 milestone for this PR?  Put the Issue number here: https://github.com/WorldHealthOrganization/app/issues/823
- [x] Provided detailed pull request description and a succinct title (consider template below for guidance).
- [x] Tested your changes, especially after any code review iterations.
- [x] Included any relevant screenshots of UI updates.
- [x] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md).
- [x] Verified all contributions are properly licensed pursuant to the [LICENSE](https://github.com/WorldHealthOrganization/app/blob/master/LICENSE) file in the root of the repository.
- [x] Verified your name is in the [content/credits.yaml](https://github.com/WorldHealthOrganization/app/blob/master/content/credits.yaml) file (if you want it to be).

## What does this PR accomplish?

Adds a "Notifications" setting to the Settings page. 

- The default case is where the user chose not to allow the permission during onboarding, on iOS they will have the system OS request the permission. 
- When a user has refused/disabled the notifications permission, this will trigger the OS's settings for the app to be open, so the user can manually toggle the permission. 

Requires content review + new translations.

Closes #936 
Closes #931 
Closes #823 
Closes #902

## Did you add any dependencies?

`notification_permissions: ^0.4.4` - to check permissions and launch settings screens
`pedantic: ^1.9.0` - to mark async calls "unawaited"

## How did you test the change?

iOS Simulator (iPhone 8, iOS 13.3)
Android Emulator (Nexus 5, API 29)

### CODE REVIEW

**Comment by kassim** [comment]
@britannio what you reckon?

**Comment by advayDev1** [comment]
Please separate the `pedantic` change.

**Comment by britannio** [comment]
Btw, notification_permissions uses a BSD license, should still be fine

**Comment by britannio** [comment]
For future reference `// ignore: unawaited_futures` can be placed above futures you don't want to await

File: client/flutter/lib/pages/settings_page.dart [hunk 1]
```diff
@@ -109,4 +124,50 @@ class _SettingsPageState extends State<SettingsPage> {
       title: S.of(context).homePagePageSliverListSettings,
     );
   }
+
+  Semantics switchItem({BuildContext context, String header, String info, bool isToggled, Function(bool) onToggle}) {
```
Comment (britannio) [inline client/flutter/lib/pages/settings_page.dart:128]:
```suggestion
  Widget switchItem({BuildContext context, String header, String info, bool isToggled, Function(bool) onToggle}) {
```

File: client/flutter/lib/pages/settings_page.dart [hunk 2]
```diff
@@ -51,53 +101,18 @@ class _SettingsPageState extends State<SettingsPage> {
...
-                      ),
-                    ),
-                  ),
-                ),
+                switchItem(
+                    context: context,
+                    header: S.of(context).homePagePageSliverListSettingsHeader1,
+                    info: S.of(context).homePagePageSliverListSettingsDataCollection,
+                    isToggled: _analyticsEnabled ?? false,
+                    onToggle: (val) => _toggleAnalytics(val)),
```
Comment (britannio) [inline client/flutter/lib/pages/settings_page.dart:109]:
```suggestion
                    onToggle: _toggleAnalytics),
```

File: client/flutter/lib/pages/settings_page.dart [hunk 3]
```diff
@@ -51,53 +101,18 @@ class _SettingsPageState extends State<SettingsPage> {
...
+                switchItem(
+                    context: context,
+                    header: S.of(context).homePagePageSliverListSettingsNotificationsHeader,
+                    info: S.of(context).homePagePageSliverListSettingsNotificationsInfo,
+                    isToggled: _notificationsEnabled ?? false,
+                    onToggle: (val) => _toggleNotifications(val)),
```
Comment (britannio) [inline client/flutter/lib/pages/settings_page.dart:115]:
```suggestion
                    onToggle: _toggleNotifications),
```

File: client/flutter/lib/pages/settings_page.dart (hunk 1 above)
Comment (britannio) [inline client/flutter/lib/pages/settings_page.dart:128]: Thanks for abstracting this.

File: client/flutter/lib/main.dart [hunk 4]
```diff
@@ -83,36 +83,8 @@ class _MyAppState extends State<MyApp> {
     // onMessage: Fires when app is foreground
```
Comment (epicfaace) [inline client/flutter/lib/main.dart:83]: Can you move these comments to where `onMessage`, `onLaunch`, and `onResume` are now defined?

**Comment by advayDev1** [comment]
@kassim plz update to master

@patniemeyer @epicfaace can we get this review finished and in?

**Comment by epicfaace** [comment]
@kassim could you re-merge master into this branch and fix conflicts?

**Comment by kassim** [comment]
@epicfaace done!

**Pull request merged**